[Feat] 게시글 예외 처리 로직 추가하기

IToriginal opened the pull request:
## 🤔 Motivation
- 게시글에 관련된 API 요청에 대한 예외처리를 구현하였습니다.

<br>

## 💡 Key Changes
- 전역 예외처리 기능이 추가됨에 따라서 Board 서비스의 예외 처리를 위해 알맞은 HttpCode를 응답하도록 클래스를 생성하였습니다.
- 게시글 투표의 기간을 설정하기 위한 컬럼인 deadLine에 과거시간은 예외로 처리할 수 있도록, @Future 어노테이션을 추가하였습니다.

<br>

## 🧐 Insufficient Content
- 게시물의 유효기간인 deadLine의 유효성 검사 테스트를 하면서 발생한 이슈입니다. 아래와 같이 GlobalExceptionHandler에 적용된 `HttpHeaders`와 `HttpStatusCode`가 적용되어 있지 않다고 에러가 발생하여 정해진 Payload의 형식으로 응답하지 않는 것을 확인했습니다. 
```java
 @ExceptionHandler(MethodArgumentNotValidException.class)
    protected ResponseEntity<Payload> handleMethodArgumentNotValidException (
            MethodArgumentNotValidException exception,
            HttpHeaders headers,
            HttpStatusCode status,
            WebRequest request
    ) { ... }
}
```
<img width="1552" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/117193889/954baf5a-8378-4a8a-b472-40d731c96daf">

- 이를 해결하기 위해 발생한 에러를 확인했는데 요청에 적합한  파라미터가 없다는 것을 나타내는 것 같아서 `HttpHeaders`와 `HttpStatusCode`를 주석처리 후 확인해보니 Payload가 적절하게 적용되는 것을 확인했습니다. 

![image](https://github.com/four-uncles/saramara-community-server/assets/117193889/48b5115e-34fd-433e-9fac-37f6420b5a39)
<img width="1016" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/117193889/495166b7-3a1a-463b-86d2-b68273295284">

- 매개변수를 삭제하지 않았을 때와 삭제했을 때의 응답 결과가 다르게 나타난 이유는 Spring MVC의 내부 동작 차이로 추측하고 있습니다.. 
- 매개변수를 삭제하지 않은 경우에는 Spring MVC가 자동으로 HTTP 응답을 생성할 때, 여러가지 추가 정보를 함께 포함시키기 때문에 불필요한 값이 응답에 포함시켜서 이와 같은 응답을 주는 것 같습니다..
- 정확한 원인을 파악하지 못해서 함께 해결하고자 공유합니다. (단순히 없애기만 하면 해결은 되나, 동작원리에 대해 잘 모르고 있기 때문에 이와 같은 원인을 잘 파악하지 못한다고 생각합니다. 🥲 때문에 이와 같은 이유에 대해 아시는 분이 있다면 함께 고민을 나눠보면 좋을 것 같습니다.)
- ----------
**주간 회의를 통해 추가된 고민 사항 공유**
- 권한에 대한 고민이였습니다. 먼저 회원을 통해서 권한이 인증되지 않은 사용자에 대해서 1차적으로 필터링이 될텐데,,, </br> 문득 `boardId`나 `commentId `그리고 `memberId`가 `long` 타입을 통해서 정수의 값이 PK 값으로 사용되었고, API를 보내게 될 때 다른 PK에 접근이 쉬운게 아닌가 하는 의문? 걱정이 들었었습니다.
- 주간 회의를 통해 UUID로 변경하면 이와 같은 문제가 해결되지 않을까? 하는 의견을 냈었는데, UUID를 사용해도 PK값이 노출되지 않는 보장이 없기 때문에 의미가 없다는 생각을 하게 되었습니다.
- 그래서, 권한에 대한 고민으로 1차적으로 당연하게 Member에서 권한을 필터링하고, 각 서비스 내에서도 권한에 대한 예외처리를 하게 된다면 조금 더 안전한 처리가 되지 않을까 고민하였습니다.
- 현재 제가 작성한 게시글 기능에는 `updateBoard` 메서드에서 DataBase에 등록된 memberId와 일치 하지 않는다면, 권한이 없다는 403 Error를 예외로 처리하도록 개발하였습니다. 이와 같은 방안으로 조금 더 안전한 코드를 작성할 수 있지 않을까 생각해봅니다..

## 👨‍👨‍👧‍👦 To Reviewers
- @langoustinee @byeongJoo05 @hb9397 
- 위의 개발 내용에 대한 코드 리뷰 및 피드백 부탁드립니다.
- 추가적으로 개선할 수 있는 부분이나 주의해야 할 사항이 있다면 알려주시면 감사하겠습니다.

IToriginal committed:
:sparkles: feat(#62): 게시글 예외 코드 정의

@Valid를 제외한 Error 상태 코드를 정의 (GlobalException에서 Valid 예외 처리)

[500 INTERNAL SERVER EXCEPTION]
- BOARD_CREATE_FAILED: 게시글 작성에 서버 문제가 발생할 경우 500 에러코드를 반환
- BOARD_UPDATE_FAILED: 게시글 수정에 서버 문제가 발생할 경우 500 에러코드를 반환
- BOARD_DELETE_FAILED: 게시글 삭제에 서버 문제가 발생할 경우 500 에러코드를 반환

[403 FORBIDDEN EXCEPTION]
- UNAUTHORIZED_TO_UPDATE_BOARD: 게시글 작성, 수정, 삭제에 연관되어 있는 회원 고유 PK가 DB에 없는 경우 사용자가 없다 판단하여, 403 에러코드를 반환

[404 NOT FOUND EXCEPTION]
- BOARD_NOT_FOUND: 게시글 조회, 수정 및 삭제에서 대상 게시글이 DB에 없는 경우, 404 에러코드를 반환

IToriginal committed:
:hammer: fix(#62): @Valid 예외 추가

deadline의 설정을 현재시간 보다 과거로 지정할 경우, 예외처리

java.validation 어노테이션
@Future: 현재 보다 미래인지 확인하는 유효성 검사

IToriginal committed:
:sparkles: feat(#62): 게시글 예외 클래스 생성

IToriginal committed:
:sparkles: feat(#62): ExceptionHandler 적용

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/exception/BoardInternalServerException.java:
500 에러는 `GlobalExceptionHandler`에서 잡고 있어서, 안쪽 도메인 패키지에서 별도로 처리할 필요는 없을 것 같습니다!

혹시나, 의도하신 바가 있다면 공유 부탁합니다!

IToriginal replied:
게시글에서 어떤 예외를 처리할까 고민해보다가 500 Error까지 고민을 했었습니다.

제가 의도한 바는 500 Error에 대한 핸들링이였는데 ... 
GlobalExceptionHandler를 병합하기 이전에 생각했었다가 놓쳤네요. 주의깊에 코드를 살피지 못했습니다.ㅠ

해당 코드는 수정해서 반영할 예정입니다!

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/service/BoardServiceImpl.java:
```diff
@@ -128,19 +140,25 @@ public Boolean updateBoard(Long boardId,
         board.setDeadLine(requestDto.getDeadLine());
```
board 객체에 수정에 대한 파라미터를 모두 받아 수정하는 메서드 1개를 추가로 두어 수정하도록 하는 것은 어떤지 고민해봐도 좋을 것 같습니다.

IToriginal replied:
음.. 제가 이해한 바가 맞는지 모르겠네요... 
먼저 현재 작성되어 있는 상태는 아래와 같습니다.
```java
board.setTitle(requestDto.getTitle());
board.setContent(requestDto.getContent());
board.setCategoryBoard(requestDto.getCategoryBoard());
board.setDeadLine(requestDto.getDeadLine());
```
제가 이해한 바는 이 코드들을 메서드로 만들어서 사용하면 어떤가라고 이해했는데 제가 이해한 것이 맞을까요?
만약에 이해한 바가 맞다면 아래와 같이 수정이 가능하다고 판단이 되는데, 아래와 같은 방법을 의도하신게 맞나요?

```java
@Override
public Boolean updateBoard(Long boardId, BoardRequestDto.UpdateRequestDto requestDto) {
        ... <동일 코드 생략>

        // 요청된 데이터로 수정할 내용 업데이트
        updateBoardRequestData(board, requestDto);

        ... <동일 코드 생략>
}

private void updateBoardRequestData(Board board, BoardRequestDto.UpdateRequestDto requestDto) {
        board.setTitle(requestDto.getTitle());
        board.setContent(requestDto.getContent());
        board.setCategoryBoard(requestDto.getCategoryBoard());
        board.setDeadLine(requestDto.getDeadLine());
}
```

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/service/BoardServiceImpl.java:
```diff
@@ -128,19 +140,25 @@ public Boolean updateBoard(Long boardId,
-        boardRepository.save(board);
-
+        Board updateBoard = boardRepository.save(board);
+        if (updateBoard == null) {
+            throw new BoardInternalServerException(BoardErrorCode.BOARD_UPDATE_FAILED);
+        }
```
서버 내부에서 예외가 발생하면 전역 예외 처리기인 `GlobalExceptionHanlder`에서 500 에러를 잡아 보낼 것 같습니다.

기본적으로 Spring Data JPA의 save() 메서드는 null을 리턴하지 않는 것으로 알고있습니다. 

<img width="654" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/59594946/515f97b6-bce1-4a1d-bc27-0893d429a82f">

찾아보니 save 메서드의 반환값은 절대 null이 되지 않는다고 합니다.

save()에서 문제가 발생하면 `DataAccessException`이 발생하게 되어 500 예외 처리가 될 것으로 보이기 때문에 null 체크를 할 필요가 없어 보입니다.

그래서 결론은 도메인 내에서 서버 내부 예외에 대한 핸들링을 할 필요는 없어 보입니다..!

IToriginal replied:
태준님 말씀대로 save 메서드는 null을 반환하지 않네요. 전역으로 관리되는 GlobalException의 500 Error를 주의 깊게 보지 못해 500 Error를 핸들링 하려다보니 오히려 불필요한 코드를 작성해버렸네요 🥲

해당 코드는 삭제하겠습니다!

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/service/BoardServiceImpl.java:
```diff
@@ -128,19 +140,25 @@ public Boolean updateBoard(Long boardId,
+
+        boardRepository.deleteById(boardId);
+        if (boardRepository.existsById(boardId)) {
+            throw new BoardInternalServerException(BoardErrorCode.BOARD_DELETE_FAILED);
```
updateBoard 메서드와 같은 맥락으로  500 에러는 `GlobalExceptionHandler`에서 처리하니 도메인 내에서는 당장 필요가 없을 것 같다고 말씀 드릴 수 있을 것 같습니다.

IToriginal replied:
의도는 boardId를 비교하여 다른 경우 false를 리턴하는 것보다 예외로 응답하게 하려고 했었습니다. 
현재 생각은 500 Error가 아닌 boardId가 다른 경우에는 권한이 없다는 403 Error로 수정할 예정입니다.

langoustinee commented:
@IToriginal 혹시 코드 수정은 얼마나 걸릴까요~~?

IToriginal commented:
> @IToriginal 혹시 코드 수정은 얼마나 걸릴까요~~?

내일 안으로 올라갈 수정 완료 계획 중입니다. 완료 되는데로 슬랙을 통해 공유 드리겠습니다!

IToriginal committed:
:hammer: fix(#62): 코드 리뷰 반영(500 Error)

- GlobalExceptionHandler에 적용되어 있어서 전역으로 500 Error를 사용하기 때문에 별도 처리가 필요없다는 의견을 반영하였습니다.

IToriginal committed:
:recycle: Refactor(#62): 코드 리뷰 반영

- BoardServiceImpl의 updateBoard 메서드 내의 내용 업데이트 로직을 Board 엔티티에 새롭게 추가한 updateBoard 메서드로 추상화했습니다.
- 캡슐화를 통해 코드 구성과 가독성이 개선되었으며, 역할의 분리가 더 잘 이루어질 것이란 기대를 합니다.
- 코드 리뷰와 슬랙을 통해 태준님이 언급해주셨던, 업데이트 작업을 트랜잭션 내에서 처리하여 데이터의 일관성을 유지하며, 작업 중에 문제가 발생할 경우 적절한 롤백이 이루어질 수 있을 것이라 기대합니다.

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/entity/Board.java:
```diff
@@ -82,4 +82,11 @@ public void setCategoryBoard(CategoryBoard categoryBoard) {
     public void setDeadLine(LocalDateTime deadLine) {
         this.deadLine = deadLine;
     }
+
+    public void updateBoard(String title, String content, CategoryBoard categoryBoard, LocalDateTime deadLine) {
+        this.title = title;
+        this.content = content;
+        this.categoryBoard = categoryBoard;
+        this.deadLine = deadLine;
+    }
```
setter 메서드를 여러 번 호출하는 것에서 엔티티 내부 메서드로 수정하도록 하도록 수정 메서드를 수정하는 점은 피드백한대로 잘 구현된 것 같습니다.

추가적으로 이를 더 개선한다면 수정 메서드에서 파라미터로 받은 **필드값을 주입할 때 바로 주입하지 않고 또 한번 별개의 메서드로 유효성 검사를 진행하여 주입**하는 것도 좋은 개선안이 될 것 같습니다.

IToriginal replied:
updateBoard 서비스단에서 RequestDto의 UpdateRequestDto를 통해서 사용하기 때문에 유효성 검사는 RequestDto의 @Valid를 통해서 한 번 검사를 마친 값들이 주입되는 것이 아닌가 하는 생각입니다.

langoustinee commented:
@IToriginal 
setter 메서드명을 setXXX()가 아닌 updateXXX()나 changeXXX() 등으로 네이밍을 변경하는 것도 좋은 개선안 중 하나가 될 것 같습니다.


```java
public void setTitle(String title) {
        this.title = title;
    }
public void setContent(String content) {
    this.content = content;
}
public void setCategoryBoard(CategoryBoard categoryBoard) {
    this.categoryBoard = categoryBoard;
}
public void setDeadLine(LocalDateTime deadLine) {
    this.deadLine = deadLine;
}
```

IToriginal commented:
> 

좋은 생각인 것 같습니다! updateXXX()나 changeXXX()등 네이밍을 변경하면 setXXX()보다 명확한 목적?이 될 것이라 생각합니다.
네이밍 부분은 계속 고려해보고 개발을 진행해 나가면서 수정해보겠습니다

IToriginal commented:
> @IToriginal setter 메서드명을 setXXX()가 아닌 updateXXX()나 changeXXX() 등으로 네이밍을 변경하는 것도 좋은 개선안 중 하나가 될 것 같습니다.
> 
> ```java
> public void setTitle(String title) {
>         this.title = title;
>     }
> public void setContent(String content) {
>     this.content = content;
> }
> public void setCategoryBoard(CategoryBoard categoryBoard) {
>     this.categoryBoard = categoryBoard;
> }
> public void setDeadLine(LocalDateTime deadLine) {
>     this.deadLine = deadLine;
> }
> ```

좋은 생각인 것 같습니다! updateXXX()나 changeXXX()등 네이밍을 변경하면 setXXX()보다 명확한 목적?이 될 것이라 생각합니다.
네이밍 부분은 계속 고려해보고 개발을 진행해 나가면서 수정해보겠습니다
